Initial cut at laying out core principles for Specification change. (#4286)
open-telemetry/opentelemetry-specification: specification/specification-principles.md

# Specification Principles

This document defines common principles that will help designers create and extend
the OpenTelemetry specification to adapt to new use cases and fix issues.

## OpenTelemetry Mission and Values

> It should be taken in context of the [overal values of OpenTelemetry](https://opentelemetry.io/community/mission/), which lays out the following core values:

Reviewer: ```suggestion
It should be taken in context of the [overall values of OpenTelemetry](https://opentelemetry.io/community/mission/), which lays out the following core values:
```

- [Telemetry should be easy](https://opentelemetry.io/community/mission/#telemetry-should-be-easy)
- [Telemetry should be universal](https://opentelemetry.io/community/mission/#telemetry-should-be-universal)
- [Telemetry should be vendor-neutral](https://opentelemetry.io/community/mission/#telemetry-should-be-vendor-neutral)
- [Telemetry should be loosely-coupled](https://opentelemetry.io/community/mission/#telemetry-should-be-loosely-coupled)
- [Telemetry should be built-in](https://opentelemetry.io/community/mission/#telemetry-should-be-built-in)

Additionally, it lays out the following core Engineering Values:

- [Compatibility](https://opentelemetry.io/community/mission/#we-value-_compatibility_)
- [Stability](https://opentelemetry.io/community/mission/#we-value-_stability_)
- [Resilience](https://opentelemetry.io/community/mission/#we-value-_resilience_)
- [Performance](https://opentelemetry.io/community/mission/#we-value-_performance_)
  - See [Specification Performance Principles](performance.md) for more details.

In addition to these core values, there are a set of things we've learned about

> Specification work and drives how we write our Specification.

Reviewer 2: ```suggestion
Specification work and drive how we write our Specification.
```

## Specification Principles

Here are the key principles of the OpenTelemetry Specification:

- [Be User Driven](#be-user-driven)
- [Be General](#be-general)
- [Be Stable](#be-stable)
- [Be Consistent](#be-consistent)
- [Be Simple](#be-stable)

Note that at times some of these principles are at odds with each other. For
example, keeping things stable may put constraints on possible simplicity. These
principles are a rubric to guide design, and form a spectrum through which we

> evaluate additions or changes to the Specification.

Reviewer 3: It's amazing how close it is to my team guidelines https://azure.github.io/azure-sdk/general_introduction.html

Someone should publish "how to design cross-language libraries" manifesto :)

Let's look at each in more detail.

### Be User Driven

The specification is useless without the ecosystem it enables. Changes should

> focs on real world use cases, and real user needs. Additioinally, changes should

Reviewer 4: ```suggestion
focus on real world use cases, and real user needs. Additioinally, changes should
```

> focus on real world use cases, and real user needs. Additioinally, changes should

Reviewer 2: ```suggestion
focus on real world use cases, and real user needs. Additionally, changes should
```

be implementable across the entire OpenTelemetry ecosystem.

This means proposals should think "end to end" not "add this one little thing".

> Additionally, projects and proposals should provide prototypes or implementations

Reviewer 4: ```suggestion
Projects and proposals should provide prototypes or implementations
```

> prior to solidification within the Specification.

Reviewer 4: ```before changes are made to the Specification.
```

We have a few simple rules of thumb regarding prototypes:

> - API/SDK changes should be prototyped in three languages.  The goal is

Reviewer 5: ```suggestion
- API/SDK changes should be prototyped in three languages. The goal is
```

> - API / SDK Changes should be prototyped in three languages.  The goal is

Reviewer 2: ```suggestion
- API/SDK changes should be prototyped in three languages.  The goal is
```

  coverage of possible API designs, not any specific language:

>   - One language should cover typed Object-Oriented ecosystems (Java, C#, etc.)

Reviewer 5: ```suggestion
  - One language should cover typed Object-Oriented ecosystems (Java, .NET, etc.)
```

>   - One language should cover dynamically typed ecosystems (Python, Javascript)

Reviewer 5: ```suggestion
  - One language should cover dynamically typed ecosystems (Python, JavaScript)
```

>   - One language should cover dynamically typed ecosystems (Python, JavaScript)

Reviewer 6: one caveat here is: Some spec changes that are primarily motivated by/required for extremely high performance scenarios (eg: Metrics Bound Instruments), may not have enough interest in all the 3 category of languages here.

Reviewer 2: +1, related to https://github.com/open-telemetry/opentelemetry-specification/pull/4286#discussion_r1833581847

Author: These aren't meant to be strict, but rules of thumb, but subjective.

Absolutely agree with the points, but these are the exceptions that we allow when we work through details and goals.  That's why these a principles and not rules.

If you feel strongly we should call out exceptions are allowed, happy to add that.  However - I do not want us to treat this document as a "MUST" type document.  This is a rubric and deviations are allowed.

>   - One language should cover dynamically typed ecosystems (Python, Javascript)

Reviewer 2: I think this is an important requirement for API changes. However, for the SDK I think this would push the bar higher and will slow us down. I don't have any problem with a higher bar though (I think as OpenTelemetry is becoming more mature with more adoption, the bar should naturally rise).

Take a concrete example, with the proposal we would need to wait for months to get this PR merged https://github.com/open-telemetry/opentelemetry-specification/pull/3041.

Author: Addressed this below.

>   - One language should cover structural ecosystems (e.g. Go, Rust)

Reviewer 5: For consistency

```suggestion
  - One language should cover structural ecosystems (Go, Rust)
```

- Protocol changes should be prototyped both on the client and the server.
- Prototypes can be unmerged Pull Requests, existing projects, etc. but must
  demonstrate the feature with confidence that the Specification of it will
  be successful.

### Be General

> The specification needs to be implemented across a wide range of languages,
> frameworks, ecosystems and communities. OpenTelemetry components should be
> allowed to provide idiomatic experiences to their users.

Reviewer 7: This doesn't feel like something about the specification and does not tell me much about how I should write a specification change. Maybe something like this is more clear?

```suggestion
The specification needs to be implementable across a wide range of languages,
frameworks, ecosystems and communities. The specification should allow 
OpenTelemetry components to provide idiomatic experiences to their users.
```

> THe Specification should focus on *what* not *how*. When describing how, use

Reviewer 7: ```suggestion
The specification should focus on *what* not *how*. When describing how, use
```

non-normative language or supplementary guidelines, like this document.

### Be Stable

> Don't. Break. Users.

Reviewer 7: I think this is not something we strictly follow. Some specification changes after stability have broken users (e.g. those that change API interfaces necessarily break users in languages like Go), and what constitutes a breaking change is a somewhat ambiguous concept. In particular, when it comes to behavior changes, there is a lot of room for interpretation.

Can we make this clearer? When we do break users in the interfaces/functions/methods signatures, what are the rules we follow? Do we have a doc that summarizes those? If we were to make a contentious behavior change (e.g. something is unclear in the spec but de facto the ecosystem relies on it), what would be the decision rules there?

Author: None of these are "strictly" follow as laid out in this paragraph at the top:

> Note that at times some of these principles are at odds with each other. For
example, keeping things stable may put constraints on possible simplicity. These
principles are a rubric to guide design, and form a spectrum through which we
evaluate additions or changes to the Specification.

These values all require subjective evaluation and reasoning.  Otherwise we'd just list them as hard requirements with normative language.  We still want the whole community to understand these values so they can do their own subjective reasoning and make appropriate trade-offs / justifications in their contributors.

Let me know if you think that paragraph needs more emphasis or gravitas.

Reviewer 7: Thanks for clarifying. I think what I would be looking for if I were to contribute to the spec would be some understanding of what the priority among these principles/rules is then. Can we e.g. order them in the list?

Another question that is left to be answered is: what happens if we drop this principle in favor of others? What are the things we still care about? For example, I assume if we break users, we still want them to be aware that we broke them and give them a smooth migration experience if needed.

Author: > Thanks for clarifying. I think what I would be looking for if I were to contribute to the spec would be some understanding of what the priority among these principles/rules is then. Can we e.g. order them in the list?

We discussed this as well.  The priority is actually flexible too.  These are things we think are important, but you need to balance them against each other, and the priority may be different depending on feature.

> Another question that is left to be answered is: what happens if we drop this principle in favor of others? What are the things we still care about? For example, I assume if we break users, we still want them to be aware that we broke them and give them a smooth migration experience if needed.

This isn't a "black and white" thing. The principle is don't break users (as best you can). We have to balance this against vulnerability fixes, bug fixes, new features, etc.  Will we break users at some point?  Likely yes.  However we should go to great lengths to avoid it, take responsibility for aiding in changes required and really do our best to find alternatives or other mechanisms where possible.  In other words - if we need to break compatibility, we go to great lengths to make this as easy as possible for users to deal with.

These values are *not* meant to be rules.  Again - if you wanted strict rules enforced naively, we can create automation to do that. These are the subjective values the TC uses when evaluating contributions *and* the same values we've seen in Spec sponsors and hope to instill in the broader community (however I'd so most folks *do* carry these values in their day to day judgement) .

Reviewer 7: This feels like a net improvement so I am resolving this comment, but I think we should aim to be even more clear on priority order. In any case, it does not need to be something to be solved now

Yes, this is a repeat of the overall OpenTelemetry mission of
[Stability](https://opentelemetry.io/community/mission/#we-value-_stability_).
That's how important stability is.

To achieve OpenTelemetry's mission of "Telemetry should be built-in", we need to
create a set of components that are safe for users to depend on. Instability

> breaks trust, and hurts the mission of being a built-in solution.

Reviewer 8: (optional)
```suggestion
breaks trust, and hurts the mission of being a solution which application
and library authors are glad to integrate with their software out of the box.
```
Perhaps this could make clearer what's mean by "a built-in solution".

When things do change, the specification (and implementation) should do the
heavy lifting to ensure seamless, smooth experience for the OpenTelemetry
ecosystem.

### Be Consistent

> Don't make users learn new concepts and behaviors for each feature they interact

Reviewer 2: Thanks for calling this out @jsuereth! This is great!

with in OpenTelemetry.  This has three sub-principles:

> - Favor fewer broadly applicable features over many narrowly defined ones.

Author: This was meant to be about adding "one little thing" to APIs until you loose both consistency, loosely-coupled and simple nature of them.  i.e. ONE well-defined simple interface over MANY little additions here or there.

I've reworded this, but also removed that aspect from values, I think it's implicit in the other discussions.

PTAL
